tools/functional-tester/etcd-tester/checks.go

```diff
@@ -65,6 +67,93 @@ func (hc *hashChecker) Check() (err error) {
 	return nil
 }
 
+type leaseChecker struct {
+	leaseStressers []Stresser
+}
+
+func newLeaseChecker(leaseStressers []Stresser) Checker { return &leaseChecker{leaseStressers} }
+
+func (lc *leaseChecker) Check() error {
+	plog.Info("lease stresser invariant check...")
+	errc := make(chan error)
+	for _, ls := range lc.leaseStressers {
+		go func(s Stresser) { errc <- lc.invariantCheck(s) }(ls)
+	}
+	var errs []error
+	for i := 0; i < len(lc.leaseStressers); i++ {
+		if err := <-errc; err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return fmt.Errorf("lease stresser encounters error: (%v)", errs[0])
+}
+
+func (lc *leaseChecker) invariantCheck(lStresser Stresser) error {
+	ls := lStresser.(*leaseStresser)
+	// check revoked leases
+	plog.Infof("revoked leases %v", ls.revokedLeases)
+	if err := leasesCheck(true, ls, ls.revokedLeases); err != nil {
+		return err
+	}
+
+	ls.revokedLeases = make(map[int64]time.Time)
+
+	// check alive leases
+	plog.Infof("alive leases %v", ls.aliveLeases)
+	if err := leasesCheck(false, ls, ls.aliveLeases); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func leasesCheck(expired bool, ls *leaseStresser, leases map[int64]time.Time) error {
+	// timed out after 10 seconds
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	for leaseID := range leases {
+		keysExpired, err := ls.hasKeysAttachedToLeaseExpired(ctx, leaseID)
+		if err != nil {
+			plog.Errorf("hasKeysAttachedToLeaseExpired error: (%v)", err)
+			return err
+		}
+		leaseExpired, err := ls.hasLeaseExpired(ctx, leaseID)
+		if err != nil {
+			plog.Errorf("hasLeaseExpired error: (%v)", err)
+			return err
+		}
+		if leaseExpired != keysExpired {
+			return fmt.Errorf("lease %v expiration mismatch (lease expired=%v, keys expired=%v)", leaseID, leaseExpired, keysExpired)
+		}
+		if leaseExpired != expired {
+			return fmt.Errorf("lease %v expected expired=%v, got %v", leaseID, expired, leaseExpired)
+		}
+	}
+	return nil
+}
+
+type compositeChecker struct {
```

Comment:
it is a variable of checker like `hashChecker` and `leaseChecker`. I think it's the best to have the checker suffix somewhere in the name.